app/src/main/java/org/jellyfin/mobile/media/MediaService.kt

```diff
@@ -0,0 +1,597 @@
+            MediaDescriptionCompat.Builder().apply {
+                setMediaId(mediaId)
+                setTitle(item.name)
+                setSubtitle(mediaSubtitle)
+                setExtras(extras)
+                setIconUri(mediaImageUri)
+            }.build()
+
+        return MediaItem(description, flag)
+    }
+
+    private inner class MediaQueueNavigator(mediaSession: MediaSessionCompat) :
+        TimelineQueueNavigator(mediaSession) {
+        override fun getMediaDescription(
+            player: Player,
+            windowIndex: Int
+        ): MediaDescriptionCompat {
+            return currentPlaylistItems[windowIndex]
+        }
+    }
+
+    private inner class MediaPlaybackPreparer(
+        private val exoPlayer: ExoPlayer,
+        private val dataSourceFactory: DataSource.Factory,
+        private val appPreferences: AppPreferences
+    ) : MediaSessionConnector.PlaybackPreparer {
+
+        override fun getSupportedPrepareActions(): Long =
+            PlaybackStateCompat.ACTION_PREPARE_FROM_MEDIA_ID or
+                PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID
+
+        override fun onPrepare(playWhenReady: Boolean) {
+        }
+
+        override fun onPrepareFromMediaId(
+            mediaId: String,
+            playWhenReady: Boolean,
+            extras: Bundle?
+        ) {
+            val shouldShuffle = mediaId.endsWith(MediaItemType.Shuffle.toString())
+
+            if (shouldShuffle) {
+                currentPlaylistItems.shuffle()
+            }
+
+            val mediaSources = currentPlaylistItems.map { item ->
+                createMediaSource(item.mediaId!!)
+            }
+
+            var mediaItemIndex = 0
+
+            if (!shouldShuffle) {
+                mediaItemIndex = currentPlaylistItems.indexOfFirst { item ->
+                    item.mediaId == mediaId
+                }
+            }
+
+            exoPlayer.prepare(ConcatenatingMediaSource(*mediaSources.toTypedArray()))
+
+            exoPlayer.seekTo(mediaItemIndex, 0)
+        }
+
+        override fun onPrepareFromSearch(
+            query: String,
+            playWhenReady: Boolean,
+            extras: Bundle?
+        ) {
+
+        }
+
+        override fun onPrepareFromUri(uri: Uri, playWhenReady: Boolean, extras: Bundle?) = Unit
+
+        override fun onCommand(
+            player: Player,
+            controlDispatcher: ControlDispatcher,
+            command: String,
+            extras: Bundle?,
+            cb: android.os.ResultReceiver?
+        ): Boolean {
+            return false
+        }
+
+        fun createMediaSource(mediaId: String): ProgressiveMediaSource {
+            val url =
+                "%s/Audio/%s/universal?UserId=%s&DeviceId=%s&MaxStreamingBitrate=140000000&Container=opus,mp3|mp3,aac,m4a,m4b|aac,flac,webma,webm,wav,ogg&TranscodingContainer=ts&TranscodingProtocol=hls&AudioCodec=aac&api_key=%s&PlaySessionId=%s&StartTimeTicks=0&EnableRedirection=true&EnableRemoteMedia=true"
```

Comment:
Can you break this up or potentially (don't know if currently possible) use the API client for this?

Reply:
Niel mentioned that too but we don't see an api client method for it. How should I break it up?

Reply:
You could split it into multiple lines for example. Also, I recommend using the Kotlin format strings, like `"string${variable}string"`.

Reply:
Thanks. Looks much better now